Errors on build samples with VS2010

OS WinXP x86, cmake 3.4.3, opencv ver 3.1.0-291-g695a430.

Errors on build opencv:
"Q:\ocv\opencv\build_x86_pre\OpenCV.sln" (целевой объект по умолчанию) (1) ->
"Q:\ocv\opencv\build_x86_pre\ALL_BUILD.vcxproj.metaproj" (целевой объект по умо
лчанию) (2) ->
"Q:\ocv\opencv\build_x86_pre\modules\line_descriptor\example_line_descriptor_kn
n_matching.vcxproj.metaproj" (целевой объект по умолчанию) (320) ->
"Q:\ocv\opencv\build_x86_pre\modules\line_descriptor\example_line_descriptor_kn
n_matching.vcxproj" (целевой объект по умолчанию) (321) ->
(Целевой объект ClCompile) ->
  ..\..\..\..\opencv_contrib\modules\line_descriptor\samples\knn_matching.cpp(1
14): error C2668: 'pow' : ambiguous call to overloaded function [Q:\ocv\opencv\
build_x86_pre\modules\line_descriptor\example_line_descriptor_knn_matching.vcxp
roj]

"Q:\ocv\opencv\build_x86_pre\OpenCV.sln" (целевой объект по умолчанию) (1) ->
"Q:\ocv\opencv\build_x86_pre\ALL_BUILD.vcxproj.metaproj" (целевой объект по умо
лчанию) (2) ->
"Q:\ocv\opencv\build_x86_pre\modules\optflow\example_optflow_optical_flow_evalu
ation.vcxproj.metaproj" (целевой объект по умолчанию) (354) ->
"Q:\ocv\opencv\build_x86_pre\modules\optflow\example_optflow_optical_flow_evalu
ation.vcxproj" (целевой объект по умолчанию) (355) ->
  ..\..\..\..\opencv_contrib\modules\optflow\samples\optical_flow_evaluation.cp
p(43): error C2065: 'NAN' : undeclared identifier [Q:\ocv\opencv\build_x86_pre\
modules\optflow\example_optflow_optical_flow_evaluation.vcxproj]
  ..\..\..\..\opencv_contrib\modules\optflow\samples\optical_flow_evaluation.cp
p(64): error C2065: 'NAN' : undeclared identifier [Q:\ocv\opencv\build_x86_pre\
modules\optflow\example_optflow_optical_flow_evaluation.vcxproj]

**Solution**
After replacement
`result += (uchar) ( bin_vector[i] * pow( 2, i ) );`
in opencv_contrib\modules\line_descriptor\samples\knn_matching.cpp(114) on
`result += (uchar) ( bin_vector[i] * pow( static_cast<float>(2), static_cast<int>(i) ) );`
and 
replace
`result.at<float>(i, j) = NAN;`
on
`result.at<float>(i, j) = std::numeric_limits<float>::quiet_NaN();` in
opencv_contrib\modules\optflow\samples\optical_flow_evaluation.cpp(43,64)
or define NAN
`#ifndef NAN
        static const unsigned long __nan[2] = {0xffffffff, 0x7fffffff};
        #define NAN (*(const float *) __nan)
    #endif`
the build was successful.

--- a/modules/line_descriptor/samples/knn_matching.cpp
+++ b/modules/line_descriptor/samples/knn_matching.cpp
@@ -111,7 +111,7 @@ uchar invertSingleBits( uchar dividend_char, int numBits )
   /* reconvert to decimal */
   uchar result = 0;
   for ( int i = (int) bin_vector.size() - 1; i >= 0; i-- )
-    result += (uchar) ( bin_vector[i] * pow( 2, i ) );
+    result += (uchar) ( bin_vector[i] * (1 << i) );
 
   return result;
 }

--- a/modules/optflow/samples/optical_flow_evaluation.cpp
+++ b/modules/optflow/samples/optical_flow_evaluation.cpp
@@ -2,6 +2,7 @@
 #include "opencv2/video.hpp"
 #include "opencv2/optflow.hpp"
 #include <fstream>
+#include <limits>
 
 using namespace std;
 using namespace cv;
@@ -40,7 +41,7 @@ static Mat endpointError( const Mat_<Point2f>& flow1, const Mat_<Point2f>& flow2
                 const Point2f diff = u1 - u2;
                 result.at<float>(i, j) = sqrt((float)diff.ddot(diff)); //distance
             } else
-                result.at<float>(i, j) = NAN;
+                result.at<float>(i, j) = std::numeric_limits<float>::quiet_NaN();
         }
     }
     return result;
@@ -61,7 +62,7 @@ static Mat angularError( const Mat_<Point2f>& flow1, const Mat_<Point2f>& flow2 
             if ( isFlowCorrect(u1) && isFlowCorrect(u2) )
                 result.at<float>(i, j) = acos((float)(u1.ddot(u2) / norm(u1) * norm(u2)));
             else
-                result.at<float>(i, j) = NAN;
+                result.at<float>(i, j) = std::numeric_limits<float>::quiet_NaN();
         }
     }
     return result;

